src/ol/renderer/canvas/canvasmaprenderer.js

```diff
@@ -92,39 +93,60 @@ ol.renderer.canvas.Map.prototype.createLayerRenderer = function(layer) {
 /**
  * @param {ol.render.EventType} type Event type.
  * @param {olx.FrameState} frameState Frame state.
+ * @param {boolean=} opt_wrapX Wrap the x-axis.
  * @private
  */
 ol.renderer.canvas.Map.prototype.dispatchComposeEvent_ =
-    function(type, frameState) {
+    function(type, frameState, opt_wrapX) {
   var map = this.getMap();
   var context = this.context_;
   if (map.hasListener(type)) {
     var extent = frameState.extent;
     var pixelRatio = frameState.pixelRatio;
     var viewState = frameState.viewState;
+    var projection = viewState.projection;
+    var projectionExtent = projection.getExtent();
     var resolution = viewState.resolution;
     var rotation = viewState.rotation;
+    var repeatReplay = (opt_wrapX && projection.isGlobal() &&
+        !ol.extent.containsExtent(projectionExtent, extent));
 
-    ol.vec.Mat4.makeTransform2D(this.transform_,
-        this.canvas_.width / 2, this.canvas_.height / 2,
-        pixelRatio / resolution, -pixelRatio / resolution,
-        -rotation,
-        -viewState.center[0], -viewState.center[1]);
+    var transform = this.getTransform(frameState);
 
     var tolerance = ol.renderer.vector.getTolerance(resolution, pixelRatio);
-    var replayGroup = new ol.render.canvas.ReplayGroup(tolerance, extent,
+    var replayGroup = new ol.render.canvas.ReplayGroup(tolerance,
+        repeatReplay ?
+            [projectionExtent[0], extent[1], projectionExtent[2], extent[3]] :
+            extent,
         resolution);
 
     var vectorContext = new ol.render.canvas.Immediate(context, pixelRatio,
-        extent, this.transform_, rotation);
+        extent, transform, rotation);
     var composeEvent = new ol.render.Event(type, map, vectorContext,
         replayGroup, frameState, context, null);
     map.dispatchEvent(composeEvent);
 
     replayGroup.finish();
     if (!replayGroup.isEmpty()) {
-      replayGroup.replay(context, pixelRatio, this.transform_,
-          rotation, {});
+      replayGroup.replay(context, pixelRatio, transform, rotation, {});
+
+      if (repeatReplay) {
+        var startX = extent[0];
+        var worldWidth = ol.extent.getWidth(projectionExtent);
+        var world = 0;
+        while (startX < projectionExtent[0]) {
+          transform = this.getTransform(frameState, worldWidth * (--world));
```

Comment:
I think it would be clearer to decrement `world` and then use the value.  The subtleties of the prefix and postfix increment and decrement operators are not worth thinking about when reading code.